driver-workload-executor/src/main/com/mongodb/workload/WorkloadExecutor.java

```diff
@@ -0,0 +1,138 @@
+ */
+
+package com.mongodb.workload;
+
+import com.mongodb.WriteConcern;
+import com.mongodb.client.JsonPoweredCrudTestHelper;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonValue;
+import org.bson.json.JsonWriterSettings;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A workload executor for the driver.  It will execute the provided workload in a loop, until the process is terminated.
+ *
+ * The following system properties are supported:
+ *
+ *  org.mongodb.test.uri: the connection string required to connect to the cluster under test
+ *  org.mongodb.workload.spec: a JSON representation of the the workload
+ *  org.mongodb.workload.output.directory: the output directory in which to write the results.json
+ */
+public class WorkloadExecutor {
+    private static volatile boolean done;
+    private static final AtomicInteger numSuccesses = new AtomicInteger();
+    private static final AtomicInteger numFailures = new AtomicInteger();
+    private static final AtomicInteger numErrors = new AtomicInteger();
+
+    /**
+     * The main entry point
+     *
+     * @param args no arguments are expected
+     */
+    public static void main(String[] args) {
+        String outputDirectory =
+                System.getProperty("org.mongodb.workload.output.directory", System.getProperty("user.dir"));
+        
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            try {
+                done = true;
+                writeResults(outputDirectory);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }));
+
+        String connectionString = System.getProperty("org.mongodb.test.uri", "mongodb://localhost");
+
+        MongoClient mongoClient = MongoClients.create(connectionString);
+
+        String workloadSpec = System.getProperty("org.mongodb.workload.spec");
+        if (workloadSpec == null) {
+            throw new IllegalArgumentException("'org.mongodb.workload.spec' system property is required");
+        }
+        BsonDocument workload = BsonDocument.parse(workloadSpec);
+
+        MongoDatabase database = mongoClient.getDatabase(workload.getString("database").getValue());
+        MongoCollection<BsonDocument> collection = database.getCollection(workload.getString("collection").getValue(),
+                BsonDocument.class);
+
+        if (workload.containsKey("testData")) {
+            List<BsonDocument> documents = new ArrayList<>();
+            for (BsonValue document : workload.getArray("testData")) {
```

Comment:
Why is it in the workload definition?

Reply:
Not sure why it was designed in this manner. I agree that it can be moved to a top-level key instead of being in the workload definition at some point in the future.

Reply:
Done